Authority Level on collections (formerly Typed collections)

dpsutton opened the pull request:
Backend work for typed collections

Features:
- added `authority_level` to Collections table. Only valid values are "official" and null|nil at the moment. (`(s/maybe (s/enum "official"))`)
- POST to `api/collections/` and PUT to `api/collections/:id` will save this new property of a collection regardless of enterprise version. Was thinking we could just not allow editing of this field in the FE. My thinking was that if a token expired or an OSS version was accidentally deployed, it would not drop data or throw errors when saving, we could just disable the benefit of this data (not show badging, prevent search boosting)
- PUT to `api/collections/:id` with `{:update_collection_subtree_authority_level true}` will apply the supplied authority_level to an entire subtree, not just direct children.
- Search will boost official collections, dashboards in official collections, and questions in official collections
- Dashboards have collection authority_level on the toplevel dashboard and on the ordered_cards
- Cards include their collection on them so the type can be accessed there

## Api responses:

<details>
<summary>
curl 'http://localhost:3000/api/dashboard/701'
</summary>

```clojure
;; some extra fields ommitted
{:ordered_cards
 ({:collection_authority_level nil  ;; each card has a collection_authority_level
   :parameter_mappings []
   :card_id 9
   :visualization_settings {}
   :dashboard_id 701
   })
 :collection_id 2825
 :name "new dashboard"
 :collection_authority_level "official"  ;; dashboard has collection_authority_level
 :id 701
 :last-edit-info
 {:id 1
  :email "dan@metabase.com"
  :first_name "dan"
  :last_name "sutton"
  :timestamp #t "2021-06-02T19:16:14.254916Z"}
 :parameters ()
 :created_at #t "2021-06-01T16:06:52.455132Z"
 :public_uuid nil
 :points_of_interest nil}
```
</details>

<details>
<summary>
curl `http://localhost:3000/api/card/759`
</summary>

```clojure
  {:description nil
   :archived false
   :collection_position nil
   :table_id 2
   :result_metadata [...]
   :creator
   {:email "dan@metabase.com"
    :first_name "dan"
    :last_login #t "2021-06-01T16:04:09.096208Z"
    :is_qbnewb false
    :is_superuser true
    :id 1
    :last_name "sutton"
    :date_joined #t "2021-03-15T20:42:55.751834Z"
    :common_name "dan sutton"}
   :can_write false
   :database_id 1
   :enable_embedding false
   :collection_id 2825
   :query_type :query
   :name "needle"
   :dashboard_count 0
   :creator_id 1
   :updated_at #t "2021-06-02T19:24:02.738638Z"
   :made_public_by_id nil
   :embedding_params nil
   :cache_ttl nil
   :dataset_query {:database 1 :query {:source-table 2} :type :query}
   :id 759
   :display :table
   :last-edit-info
   {:id 1
    :email "dan@metabase.com"
    :first_name "dan"
    :last_name "sutton"
    :timestamp #t "2021-05-12T21:44:55.911946Z"}
   :visualization_settings {:table.pivot_column "QUANTITY" :table.cell_column "SUBTOTAL"}
   :collection
   {:authority_level :official         ;; cards already have their collection so just grab the authority_level from it
    :description nil
    :archived false
    :slug "they_go_here"
    :color "#509EE3"
    :name "they go here"
    :personal_owner_id nil
    :id 2825
    :location "/"
    :namespace nil}
   :created_at #t "2021-05-12T21:44:55.896019Z"
   :public_uuid nil}
```

dpsutton committed:
Add type on collection

dpsutton committed:
Search with collection type

dpsutton committed:
Cleanup bounded heap accumulator

dpsutton committed:
Make search conditionally bump official collection items

dpsutton committed:
collection api and tests

dpsutton committed:
Put collection type onto hydrated cards on dashboards

dpsutton committed:
Move official collection type scoring into ee codebase

dpsutton committed:
ensure ee and oss agree when not official collection

dpsutton committed:
Mark Collections tree with type

dpsutton committed:
Remove unnecessary `and`s when no check for enhancements

dpsutton committed:
Tests for setting collection tree type

dpsutton committed:
Merge branch 'master' into typed-collections

dpsutton committed:
Include hydration for collection type on dashboards

dpsutton committed:
Make sure created test collections are cleaned up

dpsutton committed:
Cleanup tests, don't search on collection_type

looks for all text fields and searches the term against that. official
would bring back all official types. no bueno

dpsutton committed:
Docstring on protocol and don't shadow comparator

dpsutton committed:
Merge branch 'master' into typed-collections

dpsutton committed:
update to new ee impl var passing style

kulyk reviewed (approved):
Everything looks great from the frontend perspective! @dpsutton, thank you for describing the changes so clearly.

Should we maybe add an extra check to ensure we can't set types for users' personal collections? For sure, we'll restrict that on the frontend as well. Sorry if it's done, and I just failed to read that in Clojure

dpsutton commented:
@kulyk that's a great idea and it is not implemented yet. I'll add it now. Hopefully there is a mechanism to ensure they cannot be nested under other collections, as it would be quite difficult to ensure that it wasn't marked as such when making a whole tree marked as official.

dpsutton committed:
Collection model ensures no type change on personal collection

dpsutton committed:
Check for personal collection when updating type

model checks for personal collection and rejects in the update but
doesn't check for children of personal collection.

dpsutton committed:
Update checking for unchangeable properties of a personal collection

camsaul commented:
Relevant GH Issue: https://github.com/metabase/metabase/issues/15773

dpsutton commented:
@camsaul sorry about that. I actually haven't seen an issue for this. It was in notion and the issue wasn't linked from there. I'll add it to notion now.

dpsutton commented:
@camsaul actually that's the wrong issue. That's the metadata issue, this issue is relating to adding a type on collections.

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -8340,6 +8340,20 @@ databaseChangeLog:
+            tableName: collection
+            columns:
+              - column:
+                  name: type
```
Don't forget to "claim" your migration number in the Slack `#migrations` channel... Tim is already using 307 for BUCM.

Also to avoid the same sort of semantic/relationship/"special" confusion we have with Field types, is it worth naming this column something more specific? I'm not 100% sure what name would be... `canonicality`? I haven't been able to find what the future things beyond `regular` or `official` might go in here.

The only reason I say anything is we also have the Collection `namespace` column to distinguish regular Collections from the special `snippet` "Folders" Collections which have their own Root Collection and hierarchy. I could see myself getting confused and typing `WHERE collection.type = 'snippet'` when I should have typed `WHERE collection.namespace = 'snippet'`. I don't think I'd make that mistake with `WHERE collection.canonicality_type = 'snippet'` or whatever

dpsutton replied:
I'm a big fan of not naming this type. The only values that inhabit this at the moment are "official" and nil. It's not clear to me where it might extend to. I believe i've seen "draft" on the table. Hard to know what to name it to but I'm quite open to renaming it if we can think of something better.

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -8340,6 +8340,20 @@ databaseChangeLog:
+              - column:
+                  name: type
+                  type: varchar(255)
+                  remarks: 'Nullable column to incidate collection type'
```
People can get this information without this comment, it would be a little nicer if the comment explained _what_ a Collection type or something about the purpose of the column

camsaul commented on src/metabase/api/collection.clj:
```diff
@@ -361,15 +361,18 @@
 (api/defendpoint POST "/"
   "Create a new Collection."
-  [:as {{:keys [name color description parent_id namespace]} :body}]
+  [:as {{:keys [name color description parent_id namespace type]} :body}]
```
Another reason to use something other than `type` is so you wouldn't end up overshadowing `clojure.core/type` everywhere... either Eastwood or Bikeshed complains about that in regular `defn` forms. I guess it can't catch it here tho

camsaul commented on src/metabase/api/collection.clj:
```diff
@@ -428,23 +431,40 @@
+   archived      (s/maybe s/Bool)
+   parent_id     (s/maybe su/IntGreaterThanZero)
+   type          collection/Type
+   type_sub_tree (s/maybe s/Bool)}
```
I find this parameter name a little opaque... how about something like `update_descendant_types` or `update_subcollection_types`? "Hey API, please `update_subcollection_types` for me" seems more obvious than "Hey API, please `type_sub_tree` for me" to me. Just my 2 cents

dpsutton replied:
that's way better. I was using `type` as a verb, like "type the sub tree" but it didn't feel right to me.

camsaul commented on src/metabase/api/collection.clj:
```diff
@@ -428,23 +431,40 @@
   (let [collection-before-update (api/write-check Collection id)]
     ;; if we're trying to *archive* the Collection, make sure we're allowed to do that
     (check-allowed-to-archive-or-unarchive collection-before-update collection-updates)
+    (when (or (and (contains? collection-updates :type)
+                   (not= type (:type collection-before-update)))
+              type_sub_tree)
+      (api/check-403 (and api/*is-superuser?*
+                          ;; pre-update of model checks if the collection is a personal collection and rejects changes
+                          ;; to type, but it doesn't check if it is a sub-collection of a personal one so we add that
+                          ;; here
+                          (not (collection/is-personal-collection-or-descendant-of-one? collection-before-update)))))
```
I've actually added a new `(can-update? instance changes)` method to the permissions interface a while back... this sort of thing could be a good place to use it. We have this sort of "if this column is changing, check these permissions" code all over the place in the API endpoints. `can-update?` is supposed to be a consistent way of checking that, similar to how we pretty consistently use `can-read?` and `can-write?`...

If I had infinite refactor time I'd move all the stuff into `can-update?` and use it everywhere. But since I don't I've been trying to migrate stuff over piecemeal and use it for new stuff as it comes up.

Up to you. I'm not too bothered either way, just thought I'd mention it in case you think it makes sense

camsaul commented on src/metabase/models/collection.clj:
```diff
@@ -688,11 +692,14 @@
-     (ex-info (tru "You're not allowed to change the owner of a Personal Collection.")
-       {:status-code 400
-        :errors      {:personal_owner_id (tru "You're not allowed to change the owner of a Personal Collection.")}})))
+  (let [unchangeable {:personal_owner_id (tru "You're not allowed to change the owner of a Personal Collection.")
+                      :type              (tru "You're not allowed to change the type of a Personal Collection.")}]
+    (when-let [[k msg] (->> unchangeable
+                            (filter (fn [[k _msg]]
+                                      (api/column-will-change? k collection-before-updates collection-updates)))
+                            first)]
+      (throw
+       (ex-info msg {:status-code 400 :errors {k msg}}))))
```
Why doesn't this loop also include the checks for `location` and `archived` below?

camsaul commented on src/metabase/models/dashboard.clj:
```diff
@@ -34,20 +34,31 @@
                             [:= :card.archived nil]]] ; e.g. DashCards with no corresponding Card, e.g. text Cards
                :order-by  [[:dashcard.created_at :asc]]})))
 
+(defn collection-type
+  "Returns the collection type associated with the dashboard"
+  {:hydrate :collection_type}
+  [dashboard-or-id]
+  (-> (db/query {:select    [[:collection.type :collection_type]]
+                 :from      [[:report_dashboard :dashboard]]
+                 :left-join [[Collection :collection] [:= :collection.id :dashboard.collection_id]]
+                 :where     [:= :dashboard.id (u/the-id dashboard-or-id)]})
+      first
+      :collection_type))
```
I think you should make this a `:batched-hydrate` function. If you fetch multiple Dashboards and hydrate `:collection_type` it will have to do a DB call for each Dashboard and we'd end up with an N+1 problem. This would be simple enough to do in a single request tho.

Toucan hydration can still use the `:batched-hydrate` functions to hydrate a single object

camsaul commented on src/metabase/search/scoring.clj:
```diff
@@ -245,49 +246,73 @@
-         q))
-     (doto q
-       (.offer item)))))
+(defn- bounded-heap-accumulator
+  "A reducing function that maintains a queue of the largest items as determined by `kompare`. The queue is bounded
+  in size by `size`. Useful if you are interested in the largest `size` number of items without keeping the entire
+  collection in memory."
+  [size kompare]
+  (fn bounded-heap-acc
+    ([] (PriorityQueue. size kompare))
+    ([^PriorityQueue q]
+     (loop [acc []]
+       (if-let [x (.poll q)]
+         (recur (conj acc x))
+         acc)))
+    ([^PriorityQueue q item]
+     (if (>= (.size q) size)
+       (let [smallest (.peek q)]
+         (if (pos? (kompare item smallest))
+           (doto q
+             (.poll)
+             (.offer item))
+           q))
+       (doto q
+         (.offer item))))))
```
This seems generally useful, should it go in `metabase.util` so we could potentially use it in other places?

Maybe package it up into a function with a name like `take-sorted` (analogous to `take`) that could return either a lazy seq or transducer like `take` would make sense

dpsutton replied:
I think in haskell there's a lazy way to sort. Using the implementation here there's no good way to do this. We obviously iterate through the entire collection and we end up with a priority queue with all of the biggest items. We could lazy-seq drain that but that seems almost useless at that point. I've added tests. I think it would be a good place for test.check but I don't think we've got that laying around. The tests are very similar to invariants that test.check excels at.

camsaul commented on src/metabase/search/scoring.clj:
```diff
@@ -245,49 +246,73 @@
+                (classloader/require 'metabase-enterprise.search.scoring)
+                (some-> (resolve 'metabase-enterprise.search.scoring/ee-scoring)
+                        var-get))
+               oss-score-impl)))
```
Doesn't look like you're doing anything with that `u/prog1`

camsaul commented on test/metabase/api/collection_test.clj:
```diff
@@ -861,7 +862,18 @@
+          (mt/user-http-request :crowberto :post 200 "collection"
+                                {:name "foo", :color "#f38630", :type "official"})
+          (testing "But they have to be valid types"
+            (mt/user-http-request :crowberto :post 400 "collection"
+                                  {:name "foo", :color "#f38630", :type "no-way-this-is-valid-type"})))
+        (testing "Non-admins cannot create a collection with a type"
+          (mt/user-http-request :rasta :post 403 "collection"
+                                {:name "foo", :color "#f38630", :type "official"}))))))
```
You should consider adding an assertion for the actual API response here as well to make sure it returns the correct error in the shape we expect

camsaul commented on test/metabase/api/collection_test.clj:
```diff
@@ -878,9 +890,56 @@
+               (-> (mt/user-http-request :crowberto :put 200 (str "collection/" (u/the-id collection))
+                                         {:name "foo" :type "official"})
+                   :type))))
+      (testing "But not for personal collections"
```
I might also be worth checking the `:type` of the object in the app DB... it's totally possible to write an endpoint that returns a version of the Collection with updated `:type` that doesn't actually persist it to the DB, and this test would still pass. Better to be safe and check both

camsaul commented on test/metabase/api/search_test.clj:
```diff
@@ -52,7 +53,9 @@
 
-(def ^:private test-collection (make-result "collection test collection", :model "collection", :collection {:id true, :name true}, :updated_at false))
+(def ^:private test-collection (make-result "collection test collection", :model "collection",
+                                            :collection {:id true, :name true :type nil},
```
This is really not important but it makes me itchy to separate the first and second key-value pair with a comma but not the second and third. I'm only bringing it up because I don't want future new hires to infer that this is the idiomatic way to format stuff or people new to Clojure think there's some sort of special reason there's one comma there instead of two or zero

camsaul commented on test/metabase/models/collection_test.clj:
```diff
@@ -1206,6 +1206,13 @@
+  (testing "We are not allowed to change the type of a Personal Collection"
+    (mt/with-temp User [my-cool-user]
+      (let [personal-collection (collection/user->personal-collection my-cool-user)]
+        (is (thrown?
+             Exception
+             (db/update! Collection (u/get-id personal-collection) :type "official"))))))
```
I think `thrown-with-msg?` would be preferable here because for all we know it's failing because of an NPE or incorrect arity or something unrelated... it's nice to know that it failed because of the condition we wanted to check, and that the correct error message is returned

camsaul reviewed (approved):
Overall this looks good to me. I left a few notes on different things. You'll need to update the migration number before merging as well

dpsutton committed:
Cleanup: type collection tree, batched hydration, combine error checks

dpsutton committed:
Cleanup

dpsutton committed:
move bounded-heap accumulator to utils

dpsutton committed:
switch to test.check testing

dpsutton committed:
Bad test just to see what our CI will report

dpsutton committed:
remove purposeful failing test (was for CI)

dpsutton committed:
collection.type -> collection.authority_level

dpsutton committed:
Test the actual ee scoring impl, not the wrapped one

locally i had enhancements enabled so the tests were passing, but in
CI it was failing as it did not have enhancements enabled

dpsutton commented:
All tests are green except for the presto tests which are failing due to changes in the CI apparatus
